File: java/51.n-queens.java
/*
 * @lc app=leetcode id=51 lang=java
 *
 * [51] N-Queens
 */

// @lc code=start
class Solution {

    List<List<String>> res = new LinkedList<>();
        boolean[] r1;
        boolean[] r2;
        boolean[] r3;

    public List<List<String>> solveNQueens(int n) {
        if (n <= 0) {
            res.add(new ArrayList<>());
            return res;
        }
        r1 = new boolean[n];
        r2 = new boolean[2*n-1];
        r3 = new boolean[2*n-1];
        char[] ts = new char[n];
        for (int i = 0; i < n; ++i) ts[i] = '.';
        backtrack(0, new LinkedList<>(), n, ts);
        return res;
    }

    private void backtrack(int i, List<String> tmpList, int n, char[] ts) {
        if (i == n) {
            res.add(new ArrayList<>(tmpList));
            return;
        }
        for (int j = 0; j < n; ++j) {
            if (!r1[j] && !r2[i+j] && !r3[(j-i)+n-1]) {
                r1[j] = true;
                r2[i+j] = true;
                r3[(j-i)+n-1] = true;
                ts[j] = 'Q';
                tmpList.add(new String(ts));
                ts[j] = '.';
                backtrack(i+1, tmpList, n, ts);
                tmpList.remove(tmpList.size()-1);
                r1[j] = false;
                r2[i+j] = false;
                r3[(j-i)+n-1] = false;

            }
        }
    }



}
// @lc code=end

File: go/51.n-queens.go
/*
 * @lc app=leetcode id=51 lang=golang
 *
 * [51] N-Queens
 */

// @lc code=start
func solveNQueens(n int) [][]string {
	res := [][]string{}
	cols := make([]bool, n)
	dia1 := make([]bool, 2*n-1)
	dia2 := make([]bool, 2*n-1)
	rowtmp := []byte(strings.Repeat(".", n))
	var backtrack func(int, []string)
	backtrack = func(i int, board []string) {
		if i == n {
			tmp := make([]string, len(board))
			copy(tmp, board)
			res = append(res, tmp)
			return
		}
		for j := 0; j < n; j++ {
			if !cols[j] && !dia1[i+j] && !dia2[n-1+(j-i)] {
				cols[j], dia1[i+j], dia2[n-1+(j-i)] = true, true, true
				rowtmp[j] = 'Q'
				board = append(board, string(rowtmp))
				rowtmp[j] = '.'
				backtrack(i+1, board)
				board = board[:len(board)-1]
				cols[j], dia1[i+j], dia2[n-1+(j-i)] = false, false, false
			}
		}
	}

	backtrack(0, []string{})
	return res
}
// @lc code=end

